net/dns/nrpt_windows.go

```diff
@@ -0,0 +1,297 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dns
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/registry"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/dnsname"
+	"tailscale.com/util/winutil"
+)
+
+const (
+	dnsBaseGP     = `SOFTWARE\Policies\Microsoft\Windows NT\DNSClient`
+	nrptBaseLocal = `SYSTEM\CurrentControlSet\Services\Dnscache\Parameters\DnsPolicyConfig`
+	nrptBaseGP    = `SOFTWARE\Policies\Microsoft\Windows NT\DNSClient\DnsPolicyConfig`
+
+	nrptOverrideDNS = 0x8 // bitmask value for "use the provided override DNS resolvers"
+
+	// Apparently NRPT rules cannot handle > 50 domains.
+	nrptMaxDomainsPerRule = 50
+
+	// This is the legacy rule ID that previous versions used when we supported
+	// only a single rule. Now that we support multiple rules are required, we
+	// generate their GUIDs and store them under the Tailscale registry key.
+	nrptSingleRuleID = `{5abe529b-675b-4486-8459-25a634dacc23}`
+
+	// This is the name of the registry value we use to save Rule IDs under
+	// the Tailscale registry key.
+	nrptRuleIDValueName = `NRPTRuleIDs`
+)
+
+var (
+	libUserenv          = windows.NewLazySystemDLL("userenv.dll")
+	procRefreshPolicyEx = libUserenv.NewProc("RefreshPolicyEx")
+)
+
+const _RP_FORCE = 1
+
+// nrptRuleDatabase ensapsulates access to the Windows Name Resolution Policy
+// Table (NRPT).
+type nrptRuleDatabase struct {
+	logf      logger.Logf
+	ruleIDs   []string
+	writeAsGP bool
+	isGPDirty bool
+}
+
+func newNRPTRuleDatabase(logf logger.Logf) *nrptRuleDatabase {
+	ret := &nrptRuleDatabase{logf: logf}
+	ret.loadRuleSubkeyNames()
+	ret.initWriteAsGP()
+	// Best-effort: if our NRPT rule exists, try to delete it. Unlike
+	// per-interface configuration, NRPT rules survive the unclean
+	// termination of the Tailscale process, and depending on the
+	// rule, it may prevent us from reaching login.tailscale.com to
+	// boot up. The bootstrap resolver logic will save us, but it
+	// slows down start-up a bunch.
+	ret.DelAllRuleKeys()
+	return ret
+}
+
+func (db *nrptRuleDatabase) loadRuleSubkeyNames() {
+	result := winutil.GetRegStrings(nrptRuleIDValueName, nil)
+	if result == nil {
+		// Use the legacy rule ID if none are specified in our registry key
+		result = []string{nrptSingleRuleID}
+	}
+	db.ruleIDs = result
+}
+
+// initWriteAsGP determines which registry path should be used for writing
+// NRPT rules. If there are rules in the GP path that don't belong to us, then
+// we should use the GP path.
+func (db *nrptRuleDatabase) initWriteAsGP() {
+	var err error
+	defer func() {
+		if err != nil {
+			db.writeAsGP = false
+		}
+	}()
+
+	dnsKey, err := registry.OpenKey(registry.LOCAL_MACHINE, dnsBaseGP, registry.READ)
+	if err != nil {
+		return
+	}
+	defer dnsKey.Close()
+
+	ki, err := dnsKey.Stat()
+	if err != nil {
+		return
+	}
+
+	// If the dnsKey contains any values, then we need to use the GP key.
```

Comment:
It also looks like empty subkeys can also trigger DNS Client: https://github.com/tailscale/tailscale/issues/4607#issuecomment-1146908637

So filtering out our subkeys is irrelevant unless we're willing to delete an empty `DnsPolicyConfig`.

Reply:
AFAICT from testing, we need to delete an empty `DnsPolicyConfig` or else the policy engine still thinks there are rules, even if there aren't any. I've modified the code to incorporate that.